controllers/envoy_gateway_extension_reconciler.go

```diff
@@ -0,0 +1,269 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+
+	envoygatewayv1alpha1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/utils/ptr"
+	gatewayapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	kuadrantv1beta3 "github.com/kuadrant/kuadrant-operator/api/v1beta3"
+	"github.com/kuadrant/kuadrant-operator/pkg/common"
+	kuadrantenvoygateway "github.com/kuadrant/kuadrant-operator/pkg/envoygateway"
+	kuadrantgatewayapi "github.com/kuadrant/kuadrant-operator/pkg/library/gatewayapi"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+	"github.com/kuadrant/kuadrant-operator/pkg/wasm"
+)
+
+// envoyGatewayExtensionReconciler reconciles Envoy Gateway EnvoyExtensionPolicy custom resources
+type envoyGatewayExtensionReconciler struct {
+	client *dynamic.DynamicClient
+}
+
+func (r *envoyGatewayExtensionReconciler) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.Reconcile,
+		Events: []controller.ResourceEventMatcher{ // matches reconciliation events that change the rate limit definitions or status of rate limit policies
+			{Kind: &kuadrantv1beta1.KuadrantGroupKind},
+			{Kind: &machinery.GatewayClassGroupKind},
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &machinery.HTTPRouteGroupKind},
+			{Kind: &kuadrantv1beta3.RateLimitPolicyGroupKind},
+			{Kind: &kuadrantenvoygateway.EnvoyExtensionPolicyGroupKind},
+		},
+	}
+}
+
+func (r *envoyGatewayExtensionReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("envoyGatewayExtensionReconciler")
+
+	logger.V(1).Info("building envoy gateway extension")
+	defer logger.V(1).Info("finished building envoy gateway extension")
+
+	// build wasm plugin configs for each gateway
+	wasmConfigs, err := r.buildWasmConfigs(ctx, state)
+	if err != nil {
+		if errors.Is(err, ErrMissingStateEffectiveRateLimitPolicies) {
+			logger.V(1).Info(err.Error())
+		} else {
+			return err
+		}
+	}
+
+	// reconcile for each gateway based on the desired wasm plugin policies calculated before
+	gateways := lo.Map(topology.Targetables().Items(func(o machinery.Object) bool {
+		return o.GroupVersionKind().GroupKind() == machinery.GatewayGroupKind
```

Comment:
This is a very good question I asked myself a few times. In the end I decided not to do that for 2 reasons:
1. DRY – the decision of which gateways we build effective RateLimitPolicies for belongs to the `effectiveRateLimitPolicyReconciler`; there, there will only be an effective RateLimitPolicy that justifies creating a gateway extension for if the gateway is linked (indirectly) to the Kuadrant object.
2. (Related to the above) decoupling – to support cleaning up gateways not affected by effective RateLimitPolicies of possibly existing extension resources (EnvoyExtensionPolicy CRs in this case) decoupled from the criteria that puts gateways in scope in the first place. In fact, these are criteria that belong not even to `effectiveRateLimitPolicyReconciler`, but to the linker function between the Kuadrant CR and (indirectly) the gateways, where eventually rules could change to having a strict subset of Envoy Gateway gateways in scope for Kuadrant, including no gateway at all (instead of all gateways, as today); and cleaning up previously created extension resources would simply magically work OTB.